File: 백준/Gold/7579. 앱/앱.py
import sys

n, m = map(int, input().split())
memory = [0] + list(map(int, input().split()))
cost = [0] + list(map(int, input().split()))
limit = sum(cost)
dp = [[0] * (limit + 1) for _ in range(n + 1)]
ans = 1e9

for i in range(1, n + 1):

    for j in range(limit + 1):
        mem = memory[i]
        c = cost[i]
        if j < cost[i]:  # 현재 앱을 비활성화할만큼의 cost가 충분하지 않은 경우
            dp[i][j] = dp[i-1][j]
        else:  # 같은 cost 내에서 현재 앱을 끈 뒤의 byte와 현재 앱을 끄지 않은 뒤의 byte를 비교
            dp[i][j] = max(dp[i - 1][j - c] + mem, dp[i - 1][j])
        if dp[i][j] >= m:
            ans = min(ans, j)
print(ans)

File: 백준/Gold/7579. 앱/앱.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[] memory = new int[n + 1];
        int[] cost = new int[n + 1];
        int limit = 0;

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            memory[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            cost[i] = Integer.parseInt(st.nextToken());
            limit += cost[i]; // 총 cost 계산
        }

        int[][] dp = new int[n + 1][limit + 1];
        int ans = Integer.MAX_VALUE;

        for (int i = 1; i <= n; i++) {
            int mem = memory[i];
            int c = cost[i];
            for (int j = 0; j <= limit; j++) {
                if (j < c) {
                    dp[i][j] = dp[i - 1][j];
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j - c] + mem, dp[i - 1][j]);
                }
                if (dp[i][j] >= m) {
                    ans = Math.min(ans, j);
                }
            }
        }
        System.out.println(ans);
    }
}